include/openssl/pkcs7.h

```diff
@@ -139,11 +126,44 @@ typedef struct {
     PKCS7_ENCRYPT *encrypted;
     ASN1_TYPE *other;
   } d;
-} PKCS7;
+};
 
-// d2i_PKCS7 parses a BER-encoded, PKCS#7 signed data ContentInfo structure from
-// |len| bytes at |*inp|, as described in |d2i_SAMPLE|.
-OPENSSL_EXPORT PKCS7 *d2i_PKCS7(PKCS7 **out, const uint8_t **inp, size_t len);
+struct pkcs7_signed_st {
+  ASN1_INTEGER *version;
+  STACK_OF(X509_ALGOR) *md_algs;
+  PKCS7 *contents;
+  STACK_OF(X509) *cert;
+  STACK_OF(X509_CRL) *crl;
+  STACK_OF(PKCS7_SIGNER_INFO) *signer_info;
+};
+
+struct pkcs7_sign_envelope_st {
+    ASN1_INTEGER *version;
+    STACK_OF(PKCS7_RECIP_INFO) *recipientinfo;
+    STACK_OF(X509_ALGOR) *md_algs;
+    PKCS7_ENC_CONTENT *enc_data;
+    STACK_OF(X509) *cert;
+    STACK_OF(X509_CRL) *crl;
+    STACK_OF(PKCS7_SIGNER_INFO) *signer_info;
+};
+
+DECLARE_ASN1_FUNCTIONS(PKCS7)
+DECLARE_ASN1_FUNCTIONS(PKCS7_ISSUER_AND_SERIAL)
```

Comment:
Ran into something similar with the OCSP ASN1 functions, most consumers consuming the module only work with the top-level structures' ASN.1 functions and don't necessarily consume the ASN.1 functions of the lower-level structures.
I think we could move some of these ASN1 declarations  to `crypto/pkcs7/internal.h`. By only exposing the more necessary top level structures, we could minimize the number of APIs we expose by quite a lot.
Same thing also applies to `PKCS7_ENC_CONTENT` and maybe `PKCS7_SIGN_ENVELOPE` / `PKCS7_DIGEST`/`PKCS7_ENVELOPE`/`PKCS7_ENCRYPT`/`PKCS7_SIGNED`/?

Note: Did a quick search, Ruby doesn't directly consume the ASN.1 functions of the above mentioned structures.

Reply:
great idea. i'll move all the sub-structure function declarations into `internal.h`.

Reply:
was able to internalize all of the sub-struct ASN.1 declarations, as well as every sub-struct except for SIGNED and SIGNER_INFO. SIGNED is referenced directly in some pre-existing tests, and SIGNER_INFO is part of SIGNED's definiton.